File: backend/src/main/java/com/distributrack/service/SetupService.java
package com.distributrack.service;

import com.distributrack.dto.request.FirstAdminRequest;
import com.distributrack.dto.response.SetupStatusResponse;
import com.distributrack.dto.response.UserResponse;

/**
 * One-time system initialization. The first SUPER_ADMIN is created here —
 * and only here — and the mechanism closes forever once any user exists.
 */
public interface SetupService {

    /**
     * @return true while the users table is empty (fresh system), i.e.
     *         the first-admin setup endpoint is still usable.
     */
    SetupStatusResponse getStatus();

    /**
     * Creates the very first SUPER_ADMIN account.
     *
     * @throws RuntimeException if the system is already initialized
     *         (any user exists), or if the email/phone is taken.
     */
    UserResponse createFirstAdmin(FirstAdminRequest request);
}

File: frontend/src/services/api/setupService.ts
import { axiosInstance } from "@/services/api/axiosInstance";
import { ENDPOINTS } from "@/constants/endpoints.constants";
import type { FirstAdminPayload, SetupStatus } from "@/types/setup.types";
import type { UserProfile } from "@/types/auth.types";

/**
 * One-time first-administrator setup API client.
 *
 * Backend reference: SetupController at /api/setup. Both calls are public
 * but the backend hard-gates them to an empty users table, so after the
 * first SUPER_ADMIN exists they are inert (createFirstAdmin rejects).
 */
export const setupService = {
  /**
   * @returns whether the first-admin setup flow is still open (fresh system).
   */
  async getStatus(): Promise<SetupStatus> {
    const response = await axiosInstance.get<SetupStatus>(ENDPOINTS.SETUP.STATUS);
    return response.data;
  },

  /**
   * Creates the very first SUPER_ADMIN. Only succeeds while the users
   * table is empty; throws an ApiError otherwise.
   */
  async createFirstAdmin(payload: FirstAdminPayload): Promise<UserProfile> {
    const response = await axiosInstance.post<UserProfile>(
      ENDPOINTS.SETUP.FIRST_ADMIN,
      payload,
    );
    return response.data;
  },
};
